File: src/flutter/xilowidgets/lib/src/xdialog.dart
import 'package:flutter/material.dart';
import 'package:flet/flet.dart';

import 'package:flet/src/controls/cupertino_alert_dialog.dart';

class XDialogControl extends StatefulWidget {
  final Control? parent;
  final Control control;
  final List<Control> children;
  final bool parentDisabled;
  final bool? parentAdaptive;
  final Widget? nextChild;
  final FletControlBackend backend;

  const XDialogControl(
      {super.key,
      this.parent,
      required this.control,
      required this.children,
      required this.parentDisabled,
      required this.parentAdaptive,
      required this.nextChild,
      required this.backend});

  @override
  State<XDialogControl> createState() => _XDialogControlState();
}

class _XDialogControlState extends State<XDialogControl>
    with FletStoreMixin {
  Widget _createXDialog() {
    bool disabled = widget.control.isDisabled || widget.parentDisabled;
    bool? adaptive =
        widget.control.attrBool("adaptive") ?? widget.parentAdaptive;
    var titleCtrls =
        widget.children.where((c) => c.name == "title" && c.isVisible);
    String titleStr = widget.control.attrString("title", "")!;
    var iconCtrls =
        widget.children.where((c) => c.name == "icon" && c.isVisible);
    var contentCtrls =
        widget.children.where((c) => c.name == "content" && c.isVisible);
    var actionCtrls =
        widget.children.where((c) => c.name == "action" && c.isVisible);
    final actionsAlignment =
        parseMainAxisAlignment(widget.control.attrString("actionsAlignment"));

    if (titleCtrls.isEmpty &&
        titleStr == "" &&
        contentCtrls.isEmpty &&
        actionCtrls.isEmpty) {
      return const ErrorControl(
          "XDialog has nothing to display. Provide at minimum one of the following: title, content, actions");
    }

    var clipBehavior =
        parseClip(widget.control.attrString("clipBehavior"), Clip.none)!;

    return AlertDialog(
      title: titleCtrls.isNotEmpty
          ? createControl(widget.control, titleCtrls.first.id, disabled,
              parentAdaptive: adaptive)
          : titleStr != ""
              ? Text(titleStr)
              : null,
      titlePadding: parseEdgeInsets(widget.control, "titlePadding"),
      content: contentCtrls.isNotEmpty
          ? createControl(widget.control, contentCtrls.first.id, disabled,
              parentAdaptive: adaptive)
          : null,
      contentPadding: parseEdgeInsets(widget.control, "contentPadding",
          const EdgeInsets.fromLTRB(24.0, 20.0, 24.0, 24.0))!,
      actions: actionCtrls
          .map((c) => createControl(widget.control, c.id, disabled,
              parentAdaptive: adaptive))
          .toList(),
      actionsPadding: parseEdgeInsets(widget.control, "actionsPadding"),
      actionsAlignment: actionsAlignment,
      shape: parseOutlinedBorder(widget.control, "shape"),
      semanticLabel: widget.control.attrString("semanticsLabel"),
      insetPadding: parseEdgeInsets(widget.control, "insetPadding",
          const EdgeInsets.symmetric(horizontal: 40.0, vertical: 24.0))!,
      iconPadding: parseEdgeInsets(widget.control, "iconPadding"),
      backgroundColor: widget.control.attrColor("bgcolor", context),
      buttonPadding: parseEdgeInsets(widget.control, "actionButtonPadding"),
      surfaceTintColor: widget.control.attrColor("surfaceTintColor", context),
      shadowColor: widget.control.attrColor("shadowColor", context),
      elevation: widget.control.attrDouble("elevation"),
      clipBehavior: clipBehavior,
      icon: iconCtrls.isNotEmpty
          ? createControl(widget.control, iconCtrls.first.id, disabled,
              parentAdaptive: adaptive)
          : null,
      iconColor: widget.control.attrColor("iconColor", context),
      scrollable: widget.control.attrBool("scrollable", false)!,
      actionsOverflowButtonSpacing:
          widget.control.attrDouble("actionsOverflowButtonSpacing"),
      alignment: parseAlignment(widget.control, "alignment"),
      contentTextStyle:
          parseTextStyle(Theme.of(context), widget.control, "contentTextStyle"),
      titleTextStyle:
          parseTextStyle(Theme.of(context), widget.control, "titleTextStyle"),
    );
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("XDialog build ($hashCode): ${widget.control.id}");

    return withPagePlatform((context, platform) {
      bool? adaptive =
          widget.control.attrBool("adaptive") ?? widget.parentAdaptive;
      if (adaptive == true &&
          (platform == TargetPlatform.iOS ||
              platform == TargetPlatform.macOS)) {
        return CupertinoAlertDialogControl(
            control: widget.control,
            parentDisabled: widget.parentDisabled,
            children: widget.children,
            nextChild: widget.nextChild,
            parentAdaptive: adaptive,
            backend: widget.backend);
      }

      bool lastOpen = widget.control.state["open"] ?? false;

      debugPrint("XDialog build: ${widget.control.id}");

      var open = widget.control.attrBool("open", false)!;
      var modal = widget.control.attrBool("modal", false)!;
      var duration = widget.control.attrInt("duration", 300)!;
      var curve = widget.control.attrString("curve", "easeOutBack");
      var transitionFrom = widget.control.attrString("transitionFrom", "top-left")!;
      var offsetScale = widget.control.attrDouble("offset_scale", 100.0)!;
      var minimum_scale = widget.control.attrDouble("minimum_scale", 0.1)!;
      var maximum_scale = widget.control.attrDouble("maximum_scale", 1.0)!;
      

      debugPrint("Current open state: $lastOpen");
      debugPrint("New open state: $open");

      if (open && (open != lastOpen)) {
        var dialog = _createXDialog();
        if (dialog is ErrorControl) {
          return dialog;
        }

        // close previous dialog
        if (ModalRoute.of(context)?.isCurrent != true) {
          Navigator.of(context).pop();
        }

        widget.control.state["open"] = open;

        WidgetsBinding.instance.addPostFrameCallback((_) {
          showGeneralDialog(
              barrierDismissible: !modal,
              barrierColor: widget.control.attrColor("barrierColor", context) ?? Colors.black54,
              pageBuilder: (context, animation, secondaryAnimation) => _createXDialog(),
              context: context,
              transitionDuration: Duration(milliseconds: duration),
              transitionBuilder: (context, animation, secondaryAnimation, child) {
                final scale = Tween<double>(begin: minimum_scale, end: maximum_scale).animate(
                  CurvedAnimation(parent: animation, curve: parseCurve(curve, Curves.easeOutCubic) ?? Curves.easeOutCubic),
                );

                final offset = Tween<Offset>(
                  begin: _getOffsetFromDirection(transitionFrom),
                  end: Offset.zero,
                ).animate(animation);

                return Transform.translate(
                  offset: offset.value * offsetScale,
                  child: Transform.scale(
                    scale: scale.value,
                    child: child, // Use the child from pageBuilder
                  ),
                );
              }
              // builder: (context) => _createXDialog()
          ).then((value) {
            lastOpen = widget.control.state["open"] ?? false;
            debugPrint("Dialog should be dismissed ($hashCode): $lastOpen");
            bool shouldDismiss = lastOpen;
            widget.control.state["open"] = false;

            if (shouldDismiss) {
              widget.backend
                  .updateControlState(widget.control.id, {"open": "false"});
              widget.backend.triggerControlEvent(widget.control.id, "dismiss");
            }
          });
        });
      } else if (open != lastOpen && lastOpen) {
        Navigator.of(context).pop();
      }

      return widget.nextChild ?? const SizedBox.shrink();
    });
  }

  Offset _getOffsetFromDirection(String direction) {
    switch (direction.toLowerCase()) {
      case "top":
        return const Offset(0, -1);
      case "bottom":
        return const Offset(0, 1);
      case "left":
        return const Offset(-1, 0);
      case "right":
        return const Offset(1, 0);
      case "top-left":
        return const Offset(-1, -1);
      case "top-right":
        return const Offset(1, -1);
      case "bottom-left":
        return const Offset(-1, 1);
      case "bottom-right":
        return const Offset(1, 1);
      case "center":
        return Offset.zero;
      default:
        return Offset.zero; // fallback is center
    }
  }

  Curve? parseCurve(String? value, [Curve? defValue]) {
    switch (value?.toLowerCase()) {
      case "bouncein":
        return Curves.bounceIn;
      case "bounceinout":
        return Curves.bounceInOut;
      case "bounceout":
        return Curves.bounceOut;
      case "decelerate":
        return Curves.decelerate;
      case "ease":
        return Curves.ease;
      case "easein":
        return Curves.easeIn;
      case "easeinback":
        return Curves.easeInBack;
      case "easeincirc":
        return Curves.easeInCirc;
      case "easeincubic":
        return Curves.easeInCubic;
      case "easeinexpo":
        return Curves.easeInExpo;
      case "easeinout":
        return Curves.easeInOut;
      case "easeinoutback":
        return Curves.easeInOutBack;
      case "easeinoutcirc":
        return Curves.easeInOutCirc;
      case "easeinoutcubic":
        return Curves.easeInOutCubic;
      case "easeinoutcubicemphasized":
        return Curves.easeInOutCubicEmphasized;
      case "easeinoutexpo":
        return Curves.easeInOutExpo;
      case "easeinoutquad":
        return Curves.easeInOutQuad;
      case "easeinoutquart":
        return Curves.easeInOutQuart;
      case "easeinoutquint":
        return Curves.easeInOutQuint;
      case "easeinoutsine":
        return Curves.easeInOutSine;
      case "easeinquad":
        return Curves.easeInQuad;
      case "easeinquart":
        return Curves.easeInQuart;
      case "easeinquint":
        return Curves.easeInQuint;
      case "easeinsine":
        return Curves.easeInSine;
      case "easeintolinear":
        return Curves.easeInToLinear;
      case "easeout":
        return Curves.easeOut;
      case "easeoutback":
        return Curves.easeOutBack;
      case "easeoutcirc":
        return Curves.easeOutCirc;
      case "easeoutcubic":
        return Curves.easeOutCubic;
      case "easeoutexpo":
        return Curves.easeOutExpo;
      case "easeoutquad":
        return Curves.easeOutQuad;
      case "easeoutquart":
        return Curves.easeOutQuart;
      case "easeoutquint":
        return Curves.easeOutQuint;
      case "easeoutsine":
        return Curves.easeOutSine;
      case "elasticin":
        return Curves.elasticIn;
      case "elasticinout":
        return Curves.elasticInOut;
      case "elasticout":
        return Curves.elasticOut;
      case "fastlineartosloweasein":
        return Curves.fastLinearToSlowEaseIn;
      case "fastoutslowin":
        return Curves.fastOutSlowIn;
      case "lineartoeaseout":
        return Curves.linearToEaseOut;
      case "slowmiddle":
        return Curves.slowMiddle;
      default:
        return defValue;
    }
  }
}

File: src/xilowidgets/xdialog.py
from typing import Any, List, Optional, Union
from enum import Enum

from flet.core.adaptive_control import AdaptiveControl
from flet.core.alignment import Alignment
from flet.core.animation import AnimationCurve
from flet.core.buttons import OutlinedBorder
from flet.core.control import Control, OptionalNumber
from flet.core.ref import Ref
from flet.core.text_style import TextStyle
from flet.core.types import (
    ClipBehavior,
    ColorEnums,
    ColorValue,
    MainAxisAlignment,
    OptionalControlEventCallable,
    PaddingValue,
)


class XDialog(AdaptiveControl):
    """
    An alert dialog informs the user about situations that require acknowledgement. An alert dialog has an optional title and an optional list of actions. The title is displayed above the content and the actions are displayed below the content.

    Example:
    ```
    import flet as ft


    def main(page: ft.Page):
        page.title = "XDialog examples"
        page.horizontal_alignment = ft.CrossAxisAlignment.CENTER

        dlg = ft.XDialog(
            title=ft.Text("Hi, this is a non-modal dialog!"),
            on_dismiss=lambda e: page.add(ft.Text("Non-modal dialog dismissed")),
        )

        def handle_close(e):
            page.close(dlg_modal)
            page.add(ft.Text(f"Modal dialog closed with action: {e.control.text}"))

        dlg_modal = ft.XDialog(
            modal=True,
            title=ft.Text("Please confirm"),
            content=ft.Text("Do you really want to delete all those files?"),
            actions=[
                ft.TextButton("Yes", on_click=handle_close),
                ft.TextButton("No", on_click=handle_close),
            ],
            actions_alignment=ft.MainAxisAlignment.END,
            on_dismiss=lambda e: page.add(
                ft.Text("Modal dialog dismissed"),
            ),
        )

        page.add(
            ft.ElevatedButton("Open dialog", on_click=lambda e: page.open(dlg)),
            ft.ElevatedButton("Open modal dialog", on_click=lambda e: page.open(dlg_modal)),
        )


    ft.app(target=main)
    ```
    -----

    Online docs: https://flet.dev/docs/controls/XDialog
    """
    class LaunchDirection(Enum):
        TOP = "top"
        BOTTOM = "bottom"
        RIGHT = "right"
        LEFT = "left"
        TOP_LEFT = "top-left"
        TOP_RIGHT = "top-right"
        BOTTOM_LEFT = "bottom-left"
        BOTTOM_RIGHT = "bottom-right"
        CENTER = "center"

    def __init__(
        self,
        modal: bool = False,
        title: Optional[Union[Control, str]] = None,
        content: Optional[Control] = None,
        actions: Optional[List[Control]] = None,
        bgcolor: Optional[ColorValue] = None,
        elevation: OptionalNumber = None,
        open_duration: OptionalNumber = None,
        animation_curve: Optional[AnimationCurve] = None,
        launch_direction: Optional[LaunchDirection] = None,
        offset_scale: OptionalNumber = None,
        minimum_scale: OptionalNumber = None,
        maximum_scale: OptionalNumber = None,
        icon: Optional[Control] = None,
        open: bool = False,
        title_padding: Optional[PaddingValue] = None,
        content_padding: Optional[PaddingValue] = None,
        actions_padding: Optional[PaddingValue] = None,
        actions_alignment: Optional[MainAxisAlignment] = None,
        shape: Optional[OutlinedBorder] = None,
        inset_padding: Optional[PaddingValue] = None,
        icon_padding: Optional[PaddingValue] = None,
        action_button_padding: Optional[PaddingValue] = None,
        surface_tint_color: Optional[ColorValue] = None,
        shadow_color: Optional[ColorValue] = None,
        icon_color: Optional[ColorValue] = None,
        scrollable: Optional[bool] = None,
        actions_overflow_button_spacing: OptionalNumber = None,
        alignment: Optional[Alignment] = None,
        content_text_style: Optional[TextStyle] = None,
        title_text_style: Optional[TextStyle] = None,
        clip_behavior: Optional[ClipBehavior] = None,
        semantics_label: Optional[str] = None,
        barrier_color: Optional[ColorValue] = None,
        on_dismiss: OptionalControlEventCallable = None,
        #
        # AdaptiveControl
        #
        ref: Optional[Ref] = None,
        disabled: Optional[bool] = None,
        visible: Optional[bool] = None,
        data: Any = None,
        adaptive: Optional[bool] = None,
    ):
        Control.__init__(
            self,
            ref=ref,
            disabled=disabled,
            visible=visible,
            data=data,
        )

        AdaptiveControl.__init__(self, adaptive=adaptive)

        self.open = open
        self.bgcolor = bgcolor
        self.elevation = elevation
        self.icon = icon
        self.modal = modal
        self.title = title
        self.title_padding = title_padding
        self.content = content
        self.content_padding = content_padding
        self.actions = actions
        self.actions_padding = actions_padding
        self.actions_alignment = actions_alignment
        self.shape = shape
        self.inset_padding = inset_padding
        self.semantics_label = semantics_label
        self.on_dismiss = on_dismiss
        self.clip_behavior = clip_behavior
        self.action_button_padding = action_button_padding
        self.shadow_color = shadow_color
        self.surface_tint_color = surface_tint_color
        self.icon_padding = icon_padding
        self.icon_color = icon_color
        self.scrollable = scrollable
        self.actions_overflow_button_spacing = actions_overflow_button_spacing
        self.alignment = alignment
        self.content_text_style = content_text_style
        self.title_text_style = title_text_style
        self.barrier_color = barrier_color

        self.open_duration = open_duration
        self.animation_curve = animation_curve
        self.launch_direction = launch_direction
        self.offset_scale = offset_scale
        self.minimum_scale = minimum_scale
        self.maximum_scale = maximum_scale

    def _get_control_name(self):
        return "xdialog"

    def before_update(self):
        super().before_update()
        assert (
            self.__title or self.__content or self.__actions
        ), "XDialog has nothing to display. Provide at minimum one of the following: title, content, actions"
        self._set_attr_json("actionsPadding", self.__actions_padding)
        self._set_attr_json("contentPadding", self.__content_padding)
        self._set_attr_json("titlePadding", self.__title_padding)
        self._set_attr_json("shape", self.__shape)
        self._set_attr_json("insetPadding", self.__inset_padding)
        self._set_attr_json("iconPadding", self.__icon_padding)
        self._set_attr_json("actionButtonPadding", self.__action_button_padding)
        self._set_attr_json("alignment", self.__alignment)
        if isinstance(self.__content_text_style, TextStyle):
            self._set_attr_json("contentTextStyle", self.__content_text_style)
        if isinstance(self.__title_text_style, TextStyle):
            self._set_attr_json("titleTextStyle", self.__title_text_style)

    def _get_children(self):
        children = []
        if isinstance(self.__title, Control):
            self.__title._set_attr_internal("n", "title")
            children.append(self.__title)
        if self.__icon:
            self.__icon._set_attr_internal("n", "icon")
            children.append(self.__icon)
        if self.__content:
            self.__content._set_attr_internal("n", "content")
            children.append(self.__content)
        for action in self.__actions:
            action._set_attr_internal("n", "action")
            children.append(action)
        return children

    # open
    @property
    def open(self) -> bool:
        return self._get_attr("open", data_type="bool", def_value=False)

    @open.setter
    def open(self, value: Optional[bool]):
        self._set_attr("open", value)
    
    @property
    def open_duration(self) -> OptionalNumber:
        return self._get_attr("duration", data_type="int", def_value=250)

    @open_duration.setter
    def open_duration(self, value: OptionalNumber = 250):
        self._set_attr("duration", value)
    
    @property
    def offset_scale(self) -> OptionalNumber:
        return self._get_attr("offset_scale", data_type="float", def_value=100.0)

    @offset_scale.setter
    def offset_scale(self, value: OptionalNumber = 100.0):
        self._set_attr("offset_scale", value)

    @property
    def minimum_scale(self) -> OptionalNumber:
        return self._get_attr("minimum_scale", data_type="float", def_value=0.1)

    @minimum_scale.setter
    def minimum_scale(self, value: OptionalNumber = 0.1):
        self._set_attr("minimum_scale", value)
    
    @property
    def maximum_scale(self) -> OptionalNumber:
        return self._get_attr("maximum_scale", data_type="float", def_value=1.0)

    @maximum_scale.setter
    def maximum_scale(self, value: OptionalNumber = 1.0):
        self._set_attr("maximum_scale", value)
    
    @property
    def animation_curve(self) -> Optional[AnimationCurve]:
        return self._get_attr("animation_curve")

    @animation_curve.setter
    def animation_curve(self, value: Optional[AnimationCurve] = AnimationCurve.LINEAR):
        value = AnimationCurve.LINEAR if value is None else value
        self._set_attr("animation_curve", value.value)
    
    @property
    def launch_direction(self) -> Optional[LaunchDirection]:
        return self._get_attr("transitionFrom")

    @launch_direction.setter
    def launch_direction(self, value: Optional[LaunchDirection] = LaunchDirection.CENTER):
        value = XDialog.LaunchDirection.CENTER if value is None else value
        self._set_attr("transitionFrom", value.value)

    # bgcolor
    @property
    def bgcolor(self) -> Optional[ColorValue]:
        return self.__bgcolor

    @bgcolor.setter
    def bgcolor(self, value: Optional[ColorValue]):
        self.__bgcolor = value
        self._set_enum_attr("bgcolor", value, ColorEnums)

    # shadow_color
    @property
    def shadow_color(self) -> Optional[ColorValue]:
        return self.__shadow_color

    @shadow_color.setter
    def shadow_color(self, value: Optional[ColorValue]):
        self.__shadow_color = value
        self._set_enum_attr("shadowColor", value, ColorEnums)

    # barrier_color
    @property
    def barrier_color(self) -> Optional[ColorValue]:
        return self.__barrier_color

    @barrier_color.setter
    def barrier_color(self, value: Optional[ColorValue]):
        self.__barrier_color = value
        self._set_enum_attr("barrierColor", value, ColorEnums)

    # surface_tint_color
    @property
    def surface_tint_color(self) -> Optional[ColorValue]:
        return self.__surface_tint_color

    @surface_tint_color.setter
    def surface_tint_color(self, value: Optional[ColorValue]):
        self.__surface_tint_color = value
        self._set_enum_attr("surfaceTintColor", value, ColorEnums)

    # icon_color
    @property
    def icon_color(self) -> Optional[ColorValue]:
        return self.__icon_color

    @icon_color.setter
    def icon_color(self, value: Optional[ColorValue]):
        self.__icon_color = value
        self._set_enum_attr("iconColor", value, ColorEnums)

    # elevation
    @property
    def elevation(self) -> OptionalNumber:
        return self._get_attr("elevation", data_type="float")

    @elevation.setter
    def elevation(self, value: OptionalNumber):
        self._set_attr("elevation", value)

    # actions_overflow_button_spacing
    @property
    def actions_overflow_button_spacing(self) -> OptionalNumber:
        return self._get_attr("actionsOverflowButtonSpacing", data_type="float")

    @actions_overflow_button_spacing.setter
    def actions_overflow_button_spacing(self, value: OptionalNumber):
        self._set_attr("actionsOverflowButtonSpacing", value)

    # scrollable
    @property
    def scrollable(self) -> bool:
        return self._get_attr("scrollable", data_type="bool", def_value=False)

    @scrollable.setter
    def scrollable(self, value: Optional[bool]):
        self._set_attr("scrollable", value)

    # alignment
    @property
    def alignment(self) -> Optional[Alignment]:
        return self.__alignment

    @alignment.setter
    def alignment(self, value: Optional[Alignment]):
        self.__alignment = value

    # content_text_style
    @property
    def content_text_style(self) -> Optional[TextStyle]:
        return self.__content_text_style

    @content_text_style.setter
    def content_text_style(self, value: Optional[TextStyle]):
        self.__content_text_style = value

    # title_text_style
    @property
    def title_text_style(self) -> Optional[TextStyle]:
        return self.__title_text_style

    @title_text_style.setter
    def title_text_style(self, value: Optional[TextStyle]):
        self.__title_text_style = value

    # modal
    @property
    def modal(self) -> bool:
        return self._get_attr("modal", data_type="bool", def_value=False)

    @modal.setter
    def modal(self, value: Optional[bool]):
        self._set_attr("modal", value)

    # title
    @property
    def title(self) -> Optional[Union[Control, str]]:
        return self.__title

    @title.setter
    def title(self, value: Optional[Union[Control, str]]):
        self.__title = value
        if not isinstance(value, Control):
            self._set_attr("title", value)

    # icon
    @property
    def icon(self) -> Optional[Control]:
        return self.__icon

    @icon.setter
    def icon(self, value: Optional[Control]):
        self.__icon = value

    # title_padding
    @property
    def title_padding(self) -> Optional[PaddingValue]:
        return self.__title_padding

    @title_padding.setter
    def title_padding(self, value: Optional[PaddingValue]):
        self.__title_padding = value

    # content
    @property
    def content(self) -> Optional[Control]:
        return self.__content

    @content.setter
    def content(self, value: Optional[Control]):
        self.__content = value

    # content_padding
    @property
    def content_padding(self) -> Optional[PaddingValue]:
        return self.__content_padding

    @content_padding.setter
    def content_padding(self, value: Optional[PaddingValue]):
        self.__content_padding = value

    # actions
    @property
    def actions(self) -> List[Control]:
        return self.__actions

    @actions.setter
    def actions(self, value: Optional[List[Control]]):
        self.__actions = value if value is not None else []

    # actions_padding
    @property
    def actions_padding(self) -> Optional[PaddingValue]:
        return self.__actions_padding

    @actions_padding.setter
    def actions_padding(self, value: Optional[PaddingValue]):
        self.__actions_padding = value

    # actions_alignment
    @property
    def actions_alignment(self) -> Optional[MainAxisAlignment]:
        return self.__actions_alignment

    @actions_alignment.setter
    def actions_alignment(self, value: Optional[MainAxisAlignment]):
        self.__actions_alignment = value
        self._set_attr(
            "actionsAlignment",
            value.value if isinstance(value, MainAxisAlignment) else value,
        )

    # shape
    @property
    def shape(self) -> Optional[OutlinedBorder]:
        return self.__shape

    @shape.setter
    def shape(self, value: Optional[OutlinedBorder]):
        self.__shape = value

    # inset_padding
    @property
    def inset_padding(self) -> Optional[PaddingValue]:
        return self.__inset_padding

    @inset_padding.setter
    def inset_padding(self, value: Optional[PaddingValue]):
        self.__inset_padding = value

    # icon_padding
    @property
    def icon_padding(self) -> Optional[PaddingValue]:
        return self.__icon_padding

    @icon_padding.setter
    def icon_padding(self, value: Optional[PaddingValue]):
        self.__icon_padding = value

    # action_button_padding
    @property
    def action_button_padding(self) -> Optional[PaddingValue]:
        return self.__action_button_padding

    @action_button_padding.setter
    def action_button_padding(self, value: Optional[PaddingValue]):
        self.__action_button_padding = value

    # semantics_label
    @property
    def semantics_label(self) -> Optional[str]:
        return self._get_attr("semanticsLabel")

    @semantics_label.setter
    def semantics_label(self, value: Optional[str]):
        self._set_attr("semanticsLabel", value)

    # clip_behavior
    @property
    def clip_behavior(self) -> Optional[ClipBehavior]:
        return self.__clip_behavior

    @clip_behavior.setter
    def clip_behavior(self, value: Optional[ClipBehavior]):
        self.__clip_behavior = value
        self._set_enum_attr("clipBehavior", value, ClipBehavior)

    # on_dismiss
    @property
    def on_dismiss(self) -> OptionalControlEventCallable:
        return self._get_event_handler("dismiss")

    @on_dismiss.setter
    def on_dismiss(self, handler: OptionalControlEventCallable):
        self._add_event_handler("dismiss", handler)